README
hellanzb                                                                  README


          ;;;;            .  .
     ... :liil ...........:..:      ,._    ,._      ...................
     :   l$$$:  _.,._       _..,,._ "$$$b. "$$$b.   `_..,,._        :::
     :   $$$$.d$$$$$$L   .d$$$$$$$$L $$$$:  $$$$: .d$$$$$$$$$;      :::
     :  :$$$$P`  T$$$$: :$$$$`  7$$F:$$$$  :$$$$ :$$$$: `$$$$ __  _  |_
     :  l$$$F   :$$$$$  8$$$l"""""` l$$$l  l$$$l l$$$l   $$$L | ) /_ |_)
     :  $$$$:   l$$$$$L `4$$$bcmang;ACID$::$$$88:`4$$$bmm$$$$;.     ...
     :    ```      ```""              ```    ```    .    ```.     ..:::..
     :..............................................:   v 0 . 1 3  `:::`
                                                                     `

                 hands-free nzb downloader and post processor
--------------------------------------------------------------------------------

INSTALLING:
--------------------------------------------------
It is recommended that you install hellanzb via one of the packaging systems
(FreeBSD ports, Mac OS X darwinports). The latest port files and the python
source tarball are available from:

http://www.hellanzb.com/distfiles

To alternatively install hellanzb manually (from the source tarball):

Prerequisites:

Unix O/S               E.g.: http://www.freebsd.org, http://www.apple.com/macosx
Python >= 2.3          http://www.python.org

Twisted >= 2.0         http://www.twistedmatrix.com *and*
Twisted Web            http://www.twistedmatrix.com/projects/web
  OR
Twisted = 1.3          http://www.twistedmatrix.com

par2cmdline =~ 0.4     http://parchive.sourceforge.net
rar or unrar =~ 3.41   http://www.rarlab.com

Optionally (For MacBinary file conversion support):
macbinconv =~ 1.1.2    http://sourceforge.net/projects/macbinconv

hellanzb requires the command line utilities "par2", and "rar" (or "unrar") 
and optionally "macbinconv" to be in your shell environment's PATH.

hellanzb can either be ran from its own directory, or to install it system wide,
simply run (as root) in the hellanzb directory:

    "python setup.py install"


CONFIGURING
--------------------------------------------------
o Copy <prefix>/etc/hellanzb.conf.sample to hellanzb.conf in the same directory
  (This is automatically done for you by the ports)

  The prefix is typically:
    /usr/local (FreeBSD) or
    /opt/local (Darwin)
    or the hellanzb directory if ran locally within it
    
o Open hellanzb.conf and edit the directories (PREFIX, DEST_DIR, etc) to your
  liking. The important directories are the QUEUE directory (where you'll put
  nzb files to add to the queue) and the DEST dir (where the finished files are
  placed)

o Configure your Usenet connection with a defineServer line. There may be
  multiple defineServer lines

o Edit the defineMusicType lines to match the music decompressors installed
  on your system (any defined music decompression command line utilities must
  be available in your shell environment's PATH)

o [Optional] Edit the growler configuration to enable Mac OS X Growl
  notifications


USING
--------------------------------------------------
Run hellanzb.py

Move .nzb files into the QUEUE directory

Watch hellanzb (or go do something else) download and process your nzb
archives, leaving the finished files in the DEST dir

Also, see the 'Safe Usage' section at the bottom of this file


MISCELLANEOUS
--------------------------------------------------

o Make sure your terminal is at least 80 columns wide otherwise you may see
  download status lines scroll off the screen

o You can make hellanzb simply post process an already downloaded archive:

  "hellanzb.py -p ArchiveDirectoryName"

  In the case the archive contains password protected rars:

  "hellanzb.py -p ArchiveDirectoryName -P rarPassword"

  By default, the -p option will attempt to contact an already running hellanzb
  queue daemon via XML RPC. If the queue daemon is successfully contacted, it
  will handle the post processing work.

  If the queue daemon cannot be contacted, post processing will take place in
  the current process (and the process will exit upon completion)

  You can force -p mode to do the post processing work in the current process
  (and not attempt to contact the queue daemon) by also specifying the -L
  option:

  "hellanzb.py -Lp ArchiveDirectoryName -P rarPassword"
  
o Post processing happens automatically after archives are downloaded. In the
  case that an automatic post processor is interrupted by hellanzb exiting
  (such as from a CTRL-C), post processing will be resumed when hellanzb is
  restarted

o hellanzb logs its output to /var/tmp/hellanzb.log by default (this can be
  turned off in the configuration file)

o In the rare case that something unexpected happens, try reproducing it with
  Hellanzb.DEBUG_MODE (or hellanzb.py -d) enabled. This will produce a
  highly verbose DEBUG file that will contain any unexpected errors that were
  not picked up by the normal log file

o The Hellanzb.DEBUG_MODE log file may become rather large. It (and the normal
  log file) can be automatically trimmed and optionally rolled over via the
  Hellanzb.LOG_FILE_MAX_BYTES and Hellanzb.LOG_FILE_BACKUP_COUNT config
  variables


SAFE USAGE
--------------------------------------------------
Copyright laws and laws pertaining to patents and inventions protect original
works of authorship and inventions. Individuals who reproduce, distribute
copies, receive copies, publicly perform or display works or inventions other
than their own and without the consent of the owners or holders of rights, or
their authorized agents, in original works of authorship or inventions, may be
in violation of copyright, patent or other intellectual property infringement.
The hellanzb authors and contributors do not encourage or condone the illegal
copying of copyrighted material. This is not intended to be legal counsel or
advice. If you have any questions, consult your attorney.


hellanzb -
http://www.hellanzb.com/
KnowledgeTap Development Group


CHANGELOG
v0.14:

o Fix NZB archives containing their own NZB file (with the same name)
  hanging hellanzb

o Quiet warnings on Python 2.6.

o Now strips whitespace from NZB data, fixes slow downloads on some
  servers

o Fixed hellanzb recursively sending NNTP GROUP commands with a group
  value with extra whitespace

o Fixed download rate > 0 after downloads were finished (broken in
  0.12)

o Fixed rar password support (-P option)

o .cbr files are no longer treated as normal rars (extracted)

o Fixed defunct post processor processes on Solaris (Thanks andy at
  stormsail)

o Fixed a stall towards the end of a download situation when using
  multiple servers pools:

When one server pool couldn't establish any connections to the server
and had queued segments particular to that pool (either during
automatic failover or due to being a fill server), and the other
server pools finished their queues, the download would mistakenly be
considered 'finished'. However decoding would never finish, causing
the stall.

o New remote call: log
  (From hellanzb.py --help)

  log logLvl logMsg     Add the specified message to the HellaNZB
                        logfile. LogLvl should be INFO, WARNING, or
                        ERROR.

  (Patch by ashleywbone)

o hellanzb now looks for the hellanzb.conf file in the following
  places/order:

  local paths: ./, ./etc/, or ~/.hellanzb/
  system paths: sys.prefix, /etc, /etc/hellanzb

o Report SAXException information when an invalid NZB could not be
  parsed (noticed by TU)

o Fixed a potential lockup with multiple defineServers that occurs
  when an article does not exist on one server and failed the CRC
  check on another server (thanks jon at sd - 6, others)

o Added optional support for the optcomplete library


v0.13:

o Can now optionally store encoded article segments in memory,
  instead of disk. This can improve performance and CPU usage in
  some cases. Set CACHE_LIMIT to a non zero value to enable.
  
  Hellanzb.CACHE_LIMIT = -1 # for an unlimited sized cache, or
  Hellanzb.CACHE_LIMIT = '100MB' # for a certain size

o Added SSL NNTP support: defineServer ssl=True. (patch
  submitted by Oscar H)

o Fixed the PostProcessor on Windows, which would produce the
  error:

  exceptions.ValueError: The usePTY parameter is not supported
                         on Windows.

o Fixed download problems on Windows, which would repeatedly
  produce errors such as:

  Archive segment: 2 does not have a valid CRC/yend line!

  (thanks Ryan van Barneveld)

o Fixed total archive time reported at the end of PostProcessing
  (downloadTime was tallied as 0)

o Fixed the Hellanzb.CATEGORIZE_DEST option: it was impossible to
  disable in v0.12

o Fixed PostProcessor handling of filenames containing double quotes

o Fixed the error message:

 exceptions.TypeError: errBack() takes exactly 2 arguments (1 given)

 occurring during failed newzbin download attempts (thanks mke & JJ2)

o Fixed the following error triggered after failed yenc CRC checks:

  File "hellanzb-0.12/Hellanzb/NZBLeecher/ArticleDecoder.py", line 685,
          in assembleNZBFile
      decodedSegmentFile = open(nzbSegment.getDestination(), 'rb')
  exceptions.IOError: [Errno 2] No such file or directory: 

  This issue was introduced in v0.11.


v0.12:

o Fixed v3.newzbin.com's DirectNZB support which was broken as of
  2007-02-18

o Improved downloader performance (CPU used and max download speed),
  especially for faster connections (such as over 40mb/s)

o Added support for libnotify notifications (patch submitted by Ruben
  Fonseca)

o No longer hiding exceptions; many were previously only shown in the
  debug log file

o Now printing the 'Transferred MB in Time at Speed (Archive)' message
  when the NZB is finished downloading, as opposed to when the archive
  has finished decoding. The difference could be significant on slower
  CPUs/ and or faster connections


v0.11:

o Fixed the error:
  ImportError: cannot import name daemonize
  with Twisted 2.5

o Added the Hellanzb.CATEGORIZE_DEST option. Enabling it saves archives
  into a sub directory of DEST_DIR named after their newzbin.com
  category (when queued using the enqueuenewzbin XMLRPC call); e.g. Apps,
  Movies, Music

o enqueuenewzbin now downloads gzipped newzbin.com NZBs via the new v3
  DirectNZB interface (Thanks DBordello and Thomas Hurst)

o Added support for fillservers (by priority)

o Now requeueing segments to other servers when they fail yDecode CRC
  checks

o Added optional support for converting MacBinary files

o Now properly parsing v3.newzbin.com NZB names

o Fixed the enqueuenewzbin error:
  exceptions.AttributeError: 'module' object has no attribute 'b64encode'
  on Python 2.3

o No longer overwriting pre-existing files during rar extraction

o Fixed handling of NZB files with invalid segment number and bytes
  attributes

o Fixed .zip files in the QUEUE_DIR constantly being rescanned for NZBs

o Fixed a couple issues when antiIdle is disabled, including reconnection
  failures after connections are lost

o antiIdle disabled connections now avoid connecting at startup, and
  disconnect immediately when there is nothing left to download. i.e. they
  only are connected to the server when necessary. antiIdle disabled fill
  servers only connect when they have something to download

o Windows compatibility changes -- hellanzb has been reported to work on
  Windows as of the 0.11 changes (thanks hax)

v0.10:

o Multiple servers are now represented with different colors in the scroller
  output

o Added the ability to run an external handler script on an archive after
  post processing

o Now including duplicate files during par2 repair

o All filename paths in the config file now expand the "~" character to the
  user's home directory

o Added support for compressed (.zip) nzbs (thanks Freddie)

o Fixed the following IOError occurring on malformed NZBs that include multiple

build_util.py
#!/usr/bin/env python
"""

build_util.py - Build related functions 

(c) Copyright 2005 Philip Jenvey
[See end of file]
"""
import distutils.util, distutils.spawn, os, re, setup, shutil, sys, tarfile
try:
    from hashlib import md5
except ImportError:
    from md5 import md5

if sys.version >= '2.5':
    from hashlib import sha256

__id__ = '$Id$'

VERSION_FILENAME = './Hellanzb/__init__.py'
# rpm builds aren't very portable when built from FreeBSD machines.
# srpms are useless
#BDIST_RPM_REQUIRES = 'python >= 2.3 python-twisted pararchive rar flac shorten'

import popen2, pty, re, signal, thread, time
SPLIT_CMDLINE_ARGS_RE = re.compile(r'( |"[^"]*")')
# FIXME: Ptyopen has been deprecated by Topen (because we now used twisted all of the
# time, all processes are ran through twisted). However it is still used by the build
# scripts. they should be converted to normal popen, or the Ptyopen code should be moved
# into the build code
    
# NOTE: Ptyopen, like popens, is odd with return codes and their status. You seem to be
# able to get the correct *return code* from a Ptyopen.wait(), but you won't be able to
# get the correct WCOREDUMP *status*. This is because pty/popen run your executable via
# /bin/sh. The shell will relay the *return code* of the process back to python correctly,
# but I suspect when you check for WCOREDUMP you're checking whether or not /bin/sh core
# dumped, not your process. The solution to this is to pass the cmd as a list -- the cmd
# and its args. This tells pty/popen to run the process directly instead of via /bin/sh,
# and you get the right WCOREDUMP *status*
class Ptyopen(popen2.Popen3):
    def __init__(self, cmd, capturestderr = False, bufsize = -1):
        """ Popen3 class (isn't this actually Popen4, capturestderr = False?) that uses ptys
        instead of pipes, to allow inline reading (instead of potential i/o buffering) of
        output from the child process. It also stores the cmd its running (as a string)
        and the thread that created the object, for later use """
        import pty
        # NOTE: most of this is cutnpaste from Popen3 minus the openpty calls
        #popen2._cleanup()
        self.prettyCmd = cmd
        cmd = self.parseCmdToList(cmd)
        self.cmd = cmd
        self.threadIdent = thread.get_ident()

        p2cread, p2cwrite = pty.openpty()
        c2pread, c2pwrite = pty.openpty()
        if capturestderr:
            errout, errin = pty.openpty()
        self.pid = os.fork()
        if self.pid == 0:
            # Child
            os.dup2(p2cread, 0)
            os.dup2(c2pwrite, 1)
            if capturestderr:
                os.dup2(errin, 2)
            self._run_child(cmd)
        os.close(p2cread)
        self.tochild = os.fdopen(p2cwrite, 'w', bufsize)
        os.close(c2pwrite)
        self.fromchild = os.fdopen(c2pread, 'r', bufsize)
        if capturestderr:
            os.close(errin)
            self.childerr = os.fdopen(errout, 'r', bufsize)
        else:
            self.childerr = None

    def parseCmdToList(self, cmd):
        cleanDoubleQuotesRe = re.compile(r'^"|"$')
        args = []
        fields = SPLIT_CMDLINE_ARGS_RE.split(cmd)
        for field in fields:
            if field == '' or field == ' ':
                continue
            args.append(cleanDoubleQuotesRe.sub('', field))
        return args

    def getPid(self):
        return self.pid

    def kill(self):
        os.kill(self.pid, signal.SIGKILL)

    def poll(self):
        """Return the exit status of the child process if it has finished,
        or -1 if it hasn't finished yet."""
        if self.sts < 0:
            try:
                pid, sts = os.waitpid(self.pid, os.WNOHANG)
                if pid == self.pid:
                    self.sts = sts
            except os.error:
                pass
        return self.sts

    def wait(self):
        """Wait for and return the exit status of the child process."""
        if self.sts < 0:
            pid, sts = os.waitpid(self.pid, 0)
            if pid == self.pid:
                self.sts = sts
        return self.sts

    def readlinesAndWait(self):
        """ Read lines and wait for the process to finish. Don't read the lines too
        quickly, otherwise we could cause a deadlock with the scroller. Slow down the
        reading by pausing shortly after every read """
        output = []
        while True:
            line = self.fromchild.readline()
            if line == '': # EOF
                break
            output.append(line)

            # Somehow the scroll locks end up getting blocked unless their consumers pause
            # as short as around 1/100th of a milli every loop. You might notice this
            # delay when nzbget scrolling looks like a slightly different FPS from within
            # hellanzb than running it directly
            time.sleep(.0001)

        returnStatus = self.wait()
        return output, os.WEXITSTATUS(returnStatus)

class Ptyopen2(Ptyopen):
    """ Ptyopen = Popen3
        Ptyopen2 = Popen4
        Python was lame for naming it that way and I am just as lame
        for following suit """
    def __init__(self, cmd, bufsize = -1):
        """ Popen3 class (isn't this actually Popen4, capturestderr = False?) that uses ptys
        instead of pipes, to allow inline reading (instead of potential i/o buffering) of
        output from the child process. It also stores the cmd its running (as a string)
        and the thread that created the object, for later use """
        import pty
        #popen2._cleanup()
        self.prettyCmd = cmd
        cmd = self.parseCmdToList(cmd)
        self.cmd = cmd
        self.threadIdent = thread.get_ident()

        p2cread, p2cwrite = pty.openpty()
        c2pread, c2pwrite = pty.openpty()
        self.pid = os.fork()
        if self.pid == 0:
            # Child
            os.dup2(p2cread, 0)
            os.dup2(c2pwrite, 1)
            os.dup2(c2pwrite, 2)
            self._run_child(cmd)
        os.close(p2cread)
        self.tochild = os.fdopen(p2cwrite, 'w', bufsize)
        os.close(c2pwrite)
        self.fromchild = os.fdopen(c2pread, 'r', bufsize)

def assertIsExe(exe_list):
    """ Abort the program if none of the specified files are executable """
    if not isinstance(exe_list, (list, tuple)):
        exe_list = [exe_list]
    if exe_list:
        for exe in exe_list:
            if exe == os.path.basename(exe):
                try:
                    fullPath = distutils.spawn.find_executable(exe)
                except:
                    raise Exception(
                        'Cannot continue program, your platform does not support '
                        'searching the path for an executable and you did not supply '
                        'the full path to the %s executable.' % exe)
            else:
                fullPath = exe
            if fullPath != None and os.access(fullPath, os.X_OK):
                return fullPath
    raise Exception('Cannot continue program, required executable not found: \'' + \
                     exe + '\'')

def assertUpToDate(workingCopyDir = None):
    """ Ensure the working copy is up to date with the repository """
    if workingCopyDir == None:
        p = Ptyopen2('svn diff')
    else:
        p = Ptyopen2('svn diff ' + workingCopyDir)
        
    output, status = p.readlinesAndWait()
    
    if len(output) > 0:
        print 'Error: Cannot continue, working copy is not up to date with repository!'
        print '       Run: svn diff'
        sys.exit(1)

def branchRelease(version):
    """ branch the code base """
    fromRepository = getRepository()
    
    repository = fromRepository
    if repository[len(repository) - 1:] == '/':
        repository = repository[:len(repository) - 1]
        
    # Branch
    branchURL = repository.replace('trunk', 'tags') + '/' + version
    print 'Branching from: ' + fromRepository + ' to: ' + branchURL
    os.system('svn copy -m "Tagging new release, version: ' + version + '" . ' + branchURL)
    
    print 'Switching working copy to the new branch'
    os.system('svn switch ' + branchURL)

def buildDist():
    """ build the source and binary distributions """
    oldArg = sys.argv

    # Build source and binary distributions
    sys.argv = [ 'setup.py', 'sdist' ]
    setup.runSetup()
    
    sys.argv = [ 'setup.py', 'bdist' ]
    setup.runSetup()

    #sys.argv = [ 'setup.py', 'bdist_rpm', '--requires', BDIST_RPM_REQUIRES ]
    #setup.runSetup()
    
    sys.argv = oldArg

def buildPort(version):
    """ build a FreeBSD port """
    print 'Building new FreeBSD port'
    portStubDir = 'freebsd-port'
    
    portDestDir = 'dist/hellanzb-freebsd-port-' + version
    destDir = os.path.join(portDestDir, 'hellanzb')

    if not os.path.isdir(portDestDir):
        os.mkdir(portDestDir)
    if not os.path.isdir(destDir):
        os.mkdir(destDir)

    # replace the version
    os.system('cat ' + os.path.join(portStubDir, 'Makefile') + ' | sed s/____VERSION____/' + \
              version + '/ > ' + os.path.join(destDir, 'Makefile'))

    # copy over the port files
    shutil.copy(os.path.join(portStubDir, 'pkg-descr'), os.path.join(destDir, 'pkg-descr'))
    shutil.copy(os.path.join(portStubDir, 'pkg-plist'), os.path.join(destDir, 'pkg-plist'))
    shutil.rmtree(os.path.join(destDir, 'files'), ignore_errors=True)
    shutil.copytree(os.path.join(portStubDir, 'files'), os.path.join(destDir, 'files'))
    dotSVNDir = os.path.join(destDir, 'files', '.svn')
    if os.path.isdir(dotSVNDir):
        shutil.rmtree(dotSVNDir)

    # create a distinfo with the checksum
    distinfo = open(os.path.join(destDir, 'distinfo'), 'w')
    distinfo.write('MD5 (hellanzb-' + version + '.tar.gz) = ' +
                 md5File('dist/hellanzb-' + version + '.tar.gz') + '\n')
    distinfo.write('SHA256 (hellanzb-' + version + '.tar.gz) = ' +
                 sha256File('dist/hellanzb-' + version + '.tar.gz') + '\n')
    distinfo.write('SIZE (hellanzb-' + version + '.tar.gz) = ' +
                 str(os.path.getsize('dist/hellanzb-' + version + '.tar.gz')) + '\n')
    distinfo.close()

    dir = portDestDir[len('dist/'):]
    createTarBall('dist', dir, dir + '.tar.gz')

def buildDPort(version):
    print 'Building new Darwin port'
    portStubDir = 'darwin-dport'
    
    portDestDir = os.path.join('dist', 'hellanzb-darwin-dport-' + version)
    destDir = os.path.join(portDestDir, 'hellanzb')

    if not os.path.isdir(destDir):
        os.makedirs(destDir)

    # replace the version, and darwinports seems to require a checksum
    checksum = md5File('dist/hellanzb-' + version + '.tar.gz')
    os.system('cat ' + os.path.join(portStubDir, 'Portfile') + ' | sed s/____VERSION____/' + \
              version + '/ | ' + 'sed s/____MD5_CHECKSUM____/' + 'md5\ ' + checksum + '/ > ' + \
              os.path.join(destDir, 'Portfile'))

    dir = portDestDir[len('dist' + os.sep):]
    createTarBall('dist', dir, dir + '.tar.gz')

def bumpVersion(oldVersion):
    """ Bump the ver number. Is dumb and expects 0.0. Will bump by .1 """
    dot = oldVersion.rfind('.')
    prefix = oldVersion[0:dot + 1]
    decimal = int(oldVersion[dot + 1:])
    decimal += 1

    return prefix + str(decimal)

def createTarBall(workingDir, dirName, fileName):
    """ tar -cxvf """
    cwd = os.getcwd()
    os.chdir(workingDir)
    
    tarBall = tarfile.open(fileName, 'w:gz')
    for file in os.listdir(dirName):
        tarBall.add(os.path.join(dirName, file))
    tarBall.close()
        
    os.chdir(cwd)

def getRepository():
    """ Determine the SVN repostiory for the cwd """
    p = Ptyopen2('svn info')
    output, status = p.readlinesAndWait()
    
    for line in output:
        if len(line) > 3 and line[0:3] == 'URL':
            return line[5:].rstrip()
        
    raise Exception('Could not determine SVN repository')

def md5File(fileName):
    """ Return the md5 checksum of the specified file """
    m = md5()
    file = open(fileName)
    for line in file.readlines():
        m.update(line)
    file.close()
    return m.hexdigest()

def sha256File(fileName):
    """ Return the SHA-256 hash of the specified file """
    if sys.version >= '2.5':
        s = sha256()
        file = open(fileName)
        for line in file.readlines():
            s.update(line)
        file.close()
        return s.hexdigest()
    else:
        p = Ptyopen2('sha256 -q ' + fileName)
        output, status = p.readlinesAndWait()
        return output[0].rstrip()

def uploadToHost(version, host, dir):
    """ Upload the new build of version to the UPLOAD_HOST """
    files = []
    for file in os.listdir('dist'):
        # Upload only files for the specified version that aren't platform specific
        if file.find('-' + version + '.') > -1 and file.find(distutils.util.get_platform()) == -1:
            files.append(file)

    if len(files) == 0:
        print 'Error, could not find files to upload'

    cmd = 'scp '
    for file in files:
            cmd += 'dist/' + file + ' '
    cmd += host + ':' + dir

    # NOTE: actually, keep the old releases around. Don't break their url so soon, someone
    # could even be installing an old port
    # First, move the old release out of the way.
    #os.system('ssh ' + host + ' mv ' + dir + '/*.gz ' + dir + '/old/')
        
    os.system(cmd)

def writeVersion(newVersion, destDir = None):
    """ Write out a new version number """
    if destDir:
        versionFile = open(os.path.join(destDir, VERSION_FILENAME), 'w')
    else:
        versionFile = open(VERSION_FILENAME, 'w')
    versionFile.write('version = \'' + newVersion + '\'\n')
    versionFile.close()

"""
Copyright (c) 2005 Philip Jenvey <pjenvey@groovie.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. The name of the author or contributors may not be used to endorse or
   promote products derived from this software without specific prior
   written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.

$Id$
"""
